google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/ClientHeaderInterceptor.java

```diff
@@ -0,0 +1,90 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.data.v2.stub.metrics;
+
+import com.google.api.core.BetaApi;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.ForwardingClientCallListener.SimpleForwardingClientCallListener;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Surface GFE server-timing metric.
+ *
+ * <p>This class exports the metric from server-timing header that tracks the latency between GFE
+ * receives the first byte of a request and reads the first byte of the response. It also tracks the
+ * number of occurrences of missing server-timing header.
+ */
+@BetaApi
+public class ClientHeaderInterceptor implements ClientInterceptor {
+  private static final Logger LOGGER = Logger.getLogger(ClientHeaderInterceptor.class.getName());
+
+  public static final Metadata.Key<String> SERVER_TIMING_HEADER_KEY =
+      Metadata.Key.of("server-timing", Metadata.ASCII_STRING_MARSHALLER);
+  private static final Pattern SERVER_TIMING_HEADER_PATTERN = Pattern.compile(".*dur=(?<dur>\\d+)");
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      final MethodDescriptor<ReqT, RespT> method, final CallOptions callOptions, Channel channel) {
+    final ClientCall<ReqT, RespT> clientCall = channel.newCall(method, callOptions);
+    final String span = callOptions.getOption(HeaderTracer.SPAN_NAME_CONTEXT_KEY);
+    final HeaderTracer tracer = callOptions.getOption(HeaderTracer.HEADER_TRACER_CONTEXT_KEY);
+    return new SimpleForwardingClientCall<ReqT, RespT>(clientCall) {
+      @Override
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        super.start(
+            new SimpleForwardingClientCallListener<RespT>(responseListener) {
+              @Override
+              public void onHeaders(Metadata headers) {
+                processHeader(headers, tracer, span);
+                super.onHeaders(headers);
+              }
+            },
+            headers);
+      }
+    };
+  }
+
+  private void processHeader(Metadata headers, HeaderTracer tracer, String span) {
+    if (tracer == null) {
+      LOGGER.warning("Couldn't find HeaderTracer in call options. Skip exporting gfe metrics");
+      return;
+    }
+    if (headers.get(SERVER_TIMING_HEADER_KEY) != null) {
+      String serverTiming = headers.get(SERVER_TIMING_HEADER_KEY);
+      Matcher matcher = SERVER_TIMING_HEADER_PATTERN.matcher(serverTiming);
+      tracer.record(RpcMeasureConstants.BIGTABLE_GFE_HEADER_MISSING_COUNT, 0L, span);
```

Comment:
Do we need to record anything here?

Reply:
This logic is moved to HeaderTracer. 
I decided to keep this so if in any case gfe_latency is not recorded (for example, if the regex failed), we can check if gfe metrics were being tracked by check if gfe_header_missing_count recorded any values.